src/manifest/owned/mod.rs

```diff
@@ -0,0 +1,412 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Owned manifest containers.
+//!
+//! TODO(mcyoung): Document this module. This should be done in a followup
+//! after `manifest::container` is refactored.
+
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::mem;
+
+use zerocopy::AsBytes;
+
+use crate::crypto::rsa;
+use crate::crypto::sha256;
+use crate::io::write::StdWrite;
+use crate::io::Read as _;
+use crate::io::Write as _;
+use crate::manifest::container::HashType;
+use crate::manifest::container::Metadata;
+use crate::manifest::container::RawTocEntry;
+use crate::manifest::Error;
+use crate::manifest::ManifestType;
+use crate::protocol::wire::WireEnum;
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+pub mod pfm;
+
+/// An "owned" manifest element.
+///
+/// This trait exists to allow [`owned::Container`] to be generic over
+/// different kinds of Cerberus manifest types. In general, users should
+/// not have to implement this type.
+///
+/// `Element::ElementType` and `Element::TYPE` are analogous to the same
+/// trait items from [`Manifest`].
+///
+/// [`owned::Container`]: struct.Container.html
+/// [`Manifest`]: ../trait.Manifest.html
+#[doc(hidden)]
+pub trait Element: Sized {
+    /// A `WireEnum` representing the different kinds of valid element types
+    /// for a Manifest that Manticore understands.
+    type ElementType: WireEnum<Wire = u8>;
+
+    /// The specific value of `ManifestType` representing the type implementing
+    /// this trait.
+    const TYPE: ManifestType;
+
+    /// Returns the element type of a specific element.
+    fn element_type(&self) -> Self::ElementType;
+
+    /// Attempts to construct an `Element` of the given type from encoded
+    /// `bytes`.
+    fn from_bytes(ty: Self::ElementType, bytes: &[u8]) -> Result<Self, Error>;
+
+    /// Attempts to encode this `Element` into bytes, using the given
+    /// padding byte as "filler".
```

Comment:
My eyebrows are raised at this filler byte, I continue reading with great interest.

Reply:
Right now this exists to allow padding with `0xff` or `0x00`. Conversations with Chris indicate that this is incorrect and `0x00` is the only correct value, so this parameter will disappear in a followup.